[Frontend] Add -emit-imported-modules: listing modules imported by the current one.

huonw opened the pull request:
This is purely designed to cheaply compute dependency graphs between
modules, and thus only lists the top-level names (i.e. not submodules)
and doesn't do any form of semantic analysis.


---

The format is also just a plain text one-per-line listing of the modules; this can be optimised/changed easily.

huonw committed:
[Frontend] Add -emit-imported-modules: listing modules imported by the current one.

This is purely designed to cheaply compute dependency graphs between
modules, and thus only lists the top-level names (i.e. not submodules)
and doesn't do any form of semantic analysis.

jrose-apple commented on include/swift/Option/Options.td:
```diff
@@ -431,6 +431,13 @@ def emit_sib : Flag<["-"], "emit-sib">,
+def emit_imported_modules : Flag<["-"], "emit-imported-modules">,
+  HelpText<"Emit a list of the imported modules">, ModeOpt,
+  Flags<[FrontendOption, NoInteractiveOption, DoesNotAffectIncrementalBuild]>;
+def emit_imported_modules_path : Separate<["-"], "emit-imported-modules-path">,
```
I don't think you need this. Since -emit-imported-modules is a primary mode, you can just use -o.

huonw replied:
Fixed.

huonw committed:
[Frontend] Handle imports in the C code tied to a Swift module.

The underlying module and/or bridging header are likely to contain
module imports that we need to know about.

jrose-apple commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -2412,6 +2412,10 @@ Decl *ClangImporter::importDeclCached(const clang::NamedDecl *ClangDecl) {
 }
 
+ArrayRef<ModuleDecl::ImportedModule> ClangImporter::getImportedModules() const {
+  return Impl.ImportedHeaderExports;
```
If this is all it's going to do, I'd prefer to just ask for `getImportedHeaderModule()` from outside and not have this entry point.

jrose-apple replied:
`importer->getImportedHeaderModule()->getImportedModules(list, ModuleDecl::ImportFilter::All)`

jrose-apple commented on lib/FrontendTool/ImportedModules.cpp:
```diff
@@ -11,43 +11,99 @@
 #include "llvm/Support/FileSystem.h"
 
-bool swift::emitImportedModules(DiagnosticEngine &diags, ModuleDecl *module,
+using llvm::StringRef;
```
`#include "swift/Basic/LLVM.h"`

jrose-apple commented on lib/FrontendTool/ImportedModules.cpp:
```diff
@@ -11,43 +11,99 @@
-  for (auto *D : Decls) {
+  // Find the imports in the main Swift code.
+  llvm::SmallVector<Decl *, 32> Decls;
+  mainModule->getDisplayDecls(Decls);
```
SourceFiles already track their ImportDecls separately; it's probably faster to iterate that.

huonw replied:
If you're thinking of the `Imports` field, it seems to only be filled by name binding, which I think this pass can't run (would require being able to resolve the dependencies).

jrose-apple replied:
Whoops, you're right, of course.

jrose-apple commented on lib/FrontendTool/ImportedModules.cpp:
```diff
@@ -11,43 +11,99 @@
+    }
+    auto clangModule = underlyingModule->findUnderlyingClangModule();
+
+    findAllClangImports(clangModule, Modules);
```
Not pretty, but I guess it gets the job done.

jrose-apple commented on lib/FrontendTool/ImportedModules.cpp:
```diff
@@ -11,43 +11,99 @@
+
+  // And now look in the C code we're possibly using.
+  auto clangImporter =
+      static_cast<ClangImporter *>(Context.getClangModuleLoader());
```
Will the importer try to load overlays or anything? We need to make sure it's *not* doing that (because they might need to be rebuilt first).

huonw replied:
I don't know how to tell; any tips?

jrose-apple replied:
Hm. You could put a dummy Foundation.swiftmodule in your lib/swift/macosx/x86_64 and see if it complains about it.

huonw replied:
Hm, there's already a Foundation.swiftmodule there, but if I empty it, then only the bridging header tests fails ("malformed module file ...").

It seems like it's https://github.com/apple/swift/blob/5aa02eba56f275f1ab41e10ff10b93d866f989e9/lib/ClangImporter/ClangImporter.cpp#L1266-L1269 , although I'm not sure I understand the comment (difference between wrapper modules and adapter modules and overlays?).

huonw replied:
Also, I'm not sure I understand the problem with overlays: we only need to get the name of the imported module, so loading overlays (or not) seems like it should be unproblematic?

jrose-apple replied:
"wrapper module" = the ClangModuleUnit that contains a clang::Module
"overlay" = a Swift module with the same name as a Clang module that lives somewhere else
"adapter module" = either an overlay or the Swift half of a mixed-source framework, also a term that's been confusing everyone forever and we/I should probably just call the mixed-source framework case an overlay as well.

The code that looks for an adapter/overlay is just going through the usual module-loading path IIRC. I would expect it to find the dummy Foundation.swiftmodule and then fail to load it.

(Note that many of the tests use the "clang-importer-sdk" that we have inside the tests/Inputs/ directory, which isn't set up to be used with the real overlays. You'll want to specifically write an ad-hoc test case for this using the real SDKs, and then come up with a more principled thing to check in later.)

huonw replied:
> The code that looks for an adapter/overlay is just going through the usual module-loading path IIRC. I would expect it to find the dummy Foundation.swiftmodule and then fail to load it.

Yes, I believe that's the failure I mentioned above.

My confusion is why this behavior matters: won't either the overlay exist and be valid (in which case loading it is perfect fine), or the overlay not exist (so we don't load it, also fine)? It seems... surprising to me if something using this was to insert invalid swiftmodules into the filesystem?

jrose-apple replied:
It could exist but need to be rebuilt!

huonw replied:
Okay, I've done something for this.

Unfortunately it seems to result in `getStdlibModule` returning null, and this module is included in the list returned by `getImportedModules`. For the moment I've just skipped null modules, however it looks like an alternative would be to use `ModuleDecl::ImportFilter::Public` instead of `...::All` in the `getImportedModules` call. It's not obvious to me if this is correct.

(Also, haven't written out a test yet.)

jrose-apple replied:
I think we do need the "all" filter, though I'm not 100% sure. I also think it's okay to load the stdlib; if we've built the compiler, we've built the stdlib, just not necessarily any of the overlays.

huonw replied:
Got the standard library thing to work better, and added testing for the overlay behavior.

huonw committed:
[Test] Ensure that -emit-imported-modules doesn't load overlays.

jrose-apple commented on test/Driver/imported_modules_bridging_header.swift:
```diff
@@ -1,2 +1,7 @@
+// RUN: not %target-swiftc_driver -typecheck %s -import-objc-header %S/Inputs/imported_modules/bridging_header.h -I %S/Inputs/imported_modules 2>&1 | %FileCheck %s
```
Ah, instead of %target-swiftc_driver, please use %target-swift-build. That includes a bunch of standard options.

huonw replied:
Hm, %target-swift-build doesn't seem to exist; is %target-swift-frontend better?

huonw replied:
Oh, maybe you mean %target-build-swift?

jrose-apple replied:
Oops. Always get it wrong if I'm not looking at it.

huonw commented:
@swift-ci Please smoke test.

huonw commented:
@swift-ci Please smoke test and merge

huonw commented:
@swift-ci Please smoke test and merge

huonw commented:
@swift-ci Please smoke test and merge
